### DISCUSSION TITLE
How to reverse-parse a command (from a pre-configured `CliktCommand` to a `List<String>`)

### DISCUSSION THREAD

**Discussion by mgroth0** [ajalt/clikt#491] [Q&A]
I just discovered the library and I am very impressed! I've already begun using it and I'm enjoying it.

One thing that I am looking for right away but having trouble finding is a way to configure a `CliktCommand` at runtime and get a List<String> from it (typesafe and without any reflection tricks)

Here is how I imagine it might look:

```kotlin

class MyCommand(): CliktCommand() {
    var arg1 by argument()
    var flag1 by option().flag("-f")
    override fun run() {
        println("arg1: $arg1")
        println("flag1: $flag1")
    }
}

fun main() {
    val arguments: List<String> = MyCommand().apply { 
        arg1 = "abc"
        flag1 = true
    }.toArguments()
    /*
     Expected value of `arguments`:
     ["my-command","arg1","-f"]
    */
}

```

Though currently, it seems arguments and options can only be `val`, and I cannot find a feature that "reverse-parses" a list of strings from a pre-configured CliktCommand

**Comment by mgroth0** [comment]
I am just not sure if this type of feature already exists, or if there has been any previous discussions on it. 

If not, I can create an issue. There I could also share some more ideas for how this could be implemented.

**Comment by ajalt** [comment]
Are you asking for something like python's [shlex.split][1]? Clikt doesn't currently expose a function like that directly, but you can get something similar by making a command with [treatUnknownOptionsAsArgs][2] and no other parameters.

[1]: https://docs.python.org/3/library/shlex.html#shlex.split
[2]: https://ajalt.github.io/clikt/options/#unknown-options

**Comment by mgroth0** [comment]
No, I am not asking about anything related to parsing command line arguments. I am asking about the opposite: generating command line arguments.

I am looking for a typesafe way to configure a CliktCommand in kotlin, and then output a List<String> that would produce that command.

So from my snippet above, I would construct an instance of `MyCommand` at runtime. Then, I would configure its parameters and options in a typesafe way, right inside of kotlin. Finally, I would output a list of strings. The guarantee is that if I use `MyCommand().main(listOfStrings)` I would get the same configured `MyCommand` that I created before. Almost like serializing and deserializing a command. But the "serialized" command is just the command in a string form, and can be pasted directly into a terminal.

**Reply by ajalt** [reply]
What's your use case for this? You could implement it yourself by iterating over all the parameters with [registeredOptions][1] etc. and printing their [values][2]

[1]: https://ajalt.github.io/clikt/api/clikt/com.github.ajalt.clikt.core/-clikt-command/registered-options.html
[2]: https://ajalt.github.io/clikt/api/clikt/com.github.ajalt.clikt.parameters.options/-option-delegate/value.html

**Reply by mgroth0** [reply]
For example, say I have a CliktCommand instance inside an application. I want to configure it and execute in a subprocess with ProcessBuilder. Currently, how is the best way to do this? We can write the command as a list of raw strings and feed it into the ProcessBuilder, but this has no compile-time/type safety.

Thank you for your advice. I can try implementing it myself, and it might be sufficient for me. 

I still might recommend supporting this as a feature so that all the little details can be properly handled. CliktCommand seems to have many configurations about how to parse a command, so all this logic might need to be considered in the reverse operation as well.

**Reply by ajalt** [reply]
Parsing is fast, so I wouldn't use a subprocess for it if it's complicating your code. But if that's a hard requirement, I would pass `argv` from `main` to the process builder. Or are you trying to use Clikt to do something other than parsing a command line?

**Reply by mgroth0** [reply]
It's about modularity, and runtime performance isn't a concern in this case. I have multiple tools, and sometimes one of them wants to execute another in a subprocess as opposed to compiling everything together as a monolith.
